src/redismodule.h

```diff
@@ -623,7 +642,17 @@ REDISMODULE_API void * (*RedisModule_Calloc)(size_t nmemb, size_t size) REDISMOD
 REDISMODULE_API char * (*RedisModule_Strdup)(const char *str) REDISMODULE_ATTR;
 REDISMODULE_API int (*RedisModule_GetApi)(const char *, void *) REDISMODULE_ATTR;
 REDISMODULE_API int (*RedisModule_CreateCommand)(RedisModuleCtx *ctx, const char *name, RedisModuleCmdFunc cmdfunc, const char *strflags, int firstkey, int lastkey, int keystep) REDISMODULE_ATTR;
-REDISMODULE_API int (*RedisModule_CreateSubcommand)(RedisModuleCtx *ctx, const char *parent_name, const char *name, RedisModuleCmdFunc cmdfunc, const char *strflags, int firstkey, int lastkey, int keystep) REDISMODULE_ATTR;
+REDISMODULE_API RedisModuleCommandProxy *(*RedisModule_GetCommandProxy)(RedisModuleCtx *ctx, const char *name) REDISMODULE_ATTR;
```

Comment:
@guybe7 I'm thinking about how a module with many commands and arguments going to look like, and I wonder if we've considered also the option of making this completely declarative, i.e. the module makes a single call and provides a complex nested structure with everything.

This is of course not mutually exclusive and modules may even take that approach internally and still call these commands, but I'm asking myself if that's even possible and if it is - what would be a better API design.

Another thing that came up in different places is how we're going to handle backwards compatibility as we move on, i.e. handling of unsupported keyspecs / argument types / flags / etc. when provided by future modules compatible with future versions of Redis.

Reply:
the sub-commands will obviously need to be called separately if the module wants to provide a separate function pointer for each.
but i suppose that all the metadata about the command can be provided with some json string.
this would still be "ABI" compatible, but we'll have to:
1. add some json parser in redis core.
2. handle parts of the errors in parsing that json internally (backwards / future compatible), instead of just error one API call and let the module decide how to proceed.

other notes:
* i might be narrow minded, and just wishing to stick to what's been working so far.
* i do understand the module code is gonna look ugly, and that in any case they don't have a high incentive to use this new API.
* i don't like adding a json parser to redis core at the moment.

Reply:
We could be declarative by passing in a struct and not JSON, I think Yossi may have been suggesting that. I think JSON is useful for the normal command files since we expect other consumers, like documentation and clients, so it makes sense that it's in a portable format like JSON. Here though, it's just C talking to C (or rust or whatever). We could even consider providing something that converts JSON into a compatible C struct that just doesn't ship with Redis.

Reply:
it was decided to take the module api out of this PR, and handle in another one.